io/src/real_sense_2_grabber.cpp

```diff
@@ -0,0 +1,356 @@
+*  Copyright (c) 2010-2011, Willow Garage, Inc.
+*  Copyright (c) 2012-, Open Perception, Inc.
+*
+*  All rights reserved.
+*
+*  Redistribution and use in source and binary forms, with or without
+*  modification, are permitted provided that the following conditions
+*  are met:
+*
+*   * Redistributions of source code must retain the above copyright
+*     notice, this list of conditions and the following disclaimer.
+*   * Redistributions in binary form must reproduce the above
+*     copyright notice, this list of conditions and the following
+*     disclaimer in the documentation and/or other materials provided
+*     with the distribution.
+*   * Neither the name of the copyright holder(s) nor the names of its
+*     contributors may be used to endorse or promote products derived
+*     from this software without specific prior written permission.
+*
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*
+*/
+
+#include <pcl/io/eigen.h>
+#include <pcl/io/boost.h>
+#include <pcl/io/grabber.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+
+#include <librealsense2/rs.hpp>
+#include <pcl/io/real_sense_2_grabber.h>
+#include <pcl/common/time.h>
+
+
+namespace pcl
+{
+  RealSense2Grabber::RealSense2Grabber ( const std::string& file_name_or_serial_number, const bool repeat_playback )
+    : signal_PointXYZ ( createSignal<signal_librealsense_PointXYZ> () )
+    , signal_PointXYZI ( createSignal<signal_librealsense_PointXYZI> () )
+    , signal_PointXYZRGB ( createSignal<signal_librealsense_PointXYZRGB> () )
+    , signal_PointXYZRGBA ( createSignal<signal_librealsense_PointXYZRGBA> () )
+    , file_name_or_serial_number_ ( file_name_or_serial_number )
+    , repeat_playback_ ( repeat_playback )
+    , quit_ ( false )
+    , running_ ( false )
+    , fps_ ( 0 )
+    , device_width_ ( 424 )
+    , device_height_ ( 240 )
+    , target_fps_ ( 30 )
+  {
+  }
+
+  RealSense2Grabber::~RealSense2Grabber ()
+  {
+    stop ();
+
+    disconnect_all_slots<signal_librealsense_PointXYZ> ();
+    disconnect_all_slots<signal_librealsense_PointXYZI> ();
+    disconnect_all_slots<signal_librealsense_PointXYZRGB> ();
+    disconnect_all_slots<signal_librealsense_PointXYZRGBA> ();
+
+    thread_.join ();
```

Comment:
`thread_.join()` can actually throw, which violates the `throw`/`noexcept` qualifier.  Actually most, if not all, of the methods in this functions are not marked as not throwing. That being said, I don't think marking them as `noexcept` is the way to go.

What's the best way to proceed here? Surround it in a try-catch block? It's a case I didn't experience before but the general idea of [this](https://stackoverflow.com/a/130123) does resonate with me and with my try-catch proposal.